File: javascript/src/smells.js
import { readFileSync } from 'node:fs';
import { SMELL_THRESHOLDS } from './constants.js';
import { FUNCTION_PATTERNS } from './metrics.js';

export function detect(filepath, language) {
  let content;
  try {
    content = readFileSync(filepath, 'utf8');
  } catch (err) {
    process.stderr.write(`Warning: cannot read ${filepath}: ${err.message}\n`);
    return [];
  }

  const lines = content.split('\n');
  const functions = findFunctions(lines, language);
  const findings = [];
  const { long_function: lf, deep_nesting: dn, long_parameter_list: lp } = SMELL_THRESHOLDS;

  for (const fn of functions) {
    const { name, startLine, bodyLines, maxDepth, paramCount } = fn;

    if (bodyLines > lf.error) {
      findings.push({ type: 'long_function', severity: 'error', file: filepath, line: startLine, function_name: name, detail: `Function is ${bodyLines} lines (threshold: ${lf.error} for error)`, value: bodyLines });
    } else if (bodyLines > lf.warning) {
      findings.push({ type: 'long_function', severity: 'warning', file: filepath, line: startLine, function_name: name, detail: `Function is ${bodyLines} lines (threshold: ${lf.warning} for warning)`, value: bodyLines });
    }

    if (maxDepth > dn.error) {
      findings.push({ type: 'deep_nesting', severity: 'error', file: filepath, line: startLine, function_name: name, detail: `Nesting depth is ${maxDepth} (threshold: ${dn.error} for error)`, value: maxDepth });
    } else if (maxDepth > dn.warning) {
      findings.push({ type: 'deep_nesting', severity: 'warning', file: filepath, line: startLine, function_name: name, detail: `Nesting depth is ${maxDepth} (threshold: ${dn.warning} for warning)`, value: maxDepth });
    }

    if (paramCount > lp.error) {
      findings.push({ type: 'long_parameter_list', severity: 'error', file: filepath, line: startLine, function_name: name, detail: `Function has ${paramCount} parameters (threshold: ${lp.error} for error)`, value: paramCount });
    } else if (paramCount > lp.warning) {
      findings.push({ type: 'long_parameter_list', severity: 'warning', file: filepath, line: startLine, function_name: name, detail: `Function has ${paramCount} parameters (threshold: ${lp.warning} for warning)`, value: paramCount });
    }
  }

  return findings;
}

export function findFunctions(lines, language) {
  const fnPats = FUNCTION_PATTERNS[language] || [];
  if (!fnPats.length) return [];

  const found = [];
  const n = lines.length;

  for (let i = 0; i < n; i++) {
    const line = lines[i];
    if (!fnPats.some(p => p.test(line))) continue;

    const fnIndent = getIndent(line);
    const name = extractName(line);
    const paramCount = countParams(line, lines, i, language);

    let bodyLines = 0;
    let maxDepth = 0;
    let j = i + 1;

    while (j < n) {
      const bline = lines[j];
      const bstripped = bline.trimEnd();

      if (!bstripped) {
        bodyLines++;
        j++;
        continue;
      }

      const bIndent = getIndent(bline);
      if (bIndent <= fnIndent) break;

      bodyLines++;
      const depth = Math.floor(bIndent / 4);
      if (depth > maxDepth) maxDepth = depth;
      j++;
    }

    found.push({ name, startLine: i + 1, bodyLines, maxDepth, paramCount });
  }

  return found;
}

function getIndent(line) {
  const expanded = line.replace(/\t/g, '    ');
  return expanded.length - expanded.trimStart().length;
}

function extractName(line) {
  let m;
  if ((m = line.match(/\bdef\s+(\w+)\s*\(/))) return m[1];
  if ((m = line.match(/\bfunction\s+(\w+)\s*\(/))) return m[1];
  if ((m = line.match(/\bfunc\s+\(?(\w+)/))) return m[1];
  if ((m = line.match(/\bfn\s+(\w+)/))) return m[1];
  if ((m = line.match(/const\s+(\w+)\s*=/))) return m[1];
  if ((m = line.match(/\bfun\s+(\w+)\s*\(/))) return m[1];
  return '<anonymous>';
}

function countParams(line, lines, start, language) {
  const parenPos = line.indexOf('(');
  if (parenPos < 0) return 0;

  let paramStr = '';
  let depth = 0;
  let collecting = false;

  for (let i = start; i < Math.min(start + 10, lines.length); i++) {
    const text = i === start ? lines[i].slice(parenPos) : lines[i];
    for (const ch of text) {
      if (ch === '(') { depth++; collecting = true; }
      else if (ch === ')') {
        depth--;
        if (depth === 0 && collecting) return parseParams(paramStr, language);
      } else if (collecting && depth > 0) {
        paramStr += ch;
      }
    }
  }
  return 0;
}

function parseParams(paramStr, language) {
  paramStr = paramStr.trim();
  if (!paramStr) return 0;

  let depth = 0;
  const params = [];
  let current = '';

  for (const ch of paramStr) {
    if ('([{<'.includes(ch)) { depth++; current += ch; }
    else if (')]}>' .includes(ch)) { depth--; current += ch; }
    else if (ch === ',' && depth === 0) {
      if (current.trim()) params.push(current.trim());
      current = '';
    } else {
      current += ch;
    }
  }
  if (current.trim()) params.push(current.trim());

  if (language === 'Python') {
    return params.filter(p => {
      const base = p.split(':')[0].split('=')[0].trim().replace(/^\*+/, '');
      return base !== 'self' && base !== 'cls';
    }).length;
  }

  return params.length;
}

File: python/scanner/smells.py
import re
import sys
from .constants import SMELL_THRESHOLDS
from .metrics import FUNCTION_PATTERNS


def detect(filepath: str, language: str) -> list[dict]:
    try:
        with open(filepath, 'r', encoding='utf-8', errors='replace') as f:
            lines = f.readlines()
    except OSError as exc:
        print(f"Warning: cannot read {filepath}: {exc}", file=sys.stderr)
        return []

    functions = _find_functions(lines, language)
    findings = []
    lf = SMELL_THRESHOLDS['long_function']
    dn = SMELL_THRESHOLDS['deep_nesting']
    lp = SMELL_THRESHOLDS['long_parameter_list']

    for fn in functions:
        name = fn['name']
        line = fn['start_line']
        body = fn['body_lines']
        depth = fn['max_depth']
        params = fn['param_count']

        if body > lf['error']:
            findings.append({
                'type': 'long_function', 'severity': 'error',
                'file': filepath, 'line': line, 'function_name': name,
                'detail': f'Function is {body} lines (threshold: {lf["error"]} for error)',
                'value': body,
            })
        elif body > lf['warning']:
            findings.append({
                'type': 'long_function', 'severity': 'warning',
                'file': filepath, 'line': line, 'function_name': name,
                'detail': f'Function is {body} lines (threshold: {lf["warning"]} for warning)',
                'value': body,
            })

        if depth > dn['error']:
            findings.append({
                'type': 'deep_nesting', 'severity': 'error',
                'file': filepath, 'line': line, 'function_name': name,
                'detail': f'Nesting depth is {depth} (threshold: {dn["error"]} for error)',
                'value': depth,
            })
        elif depth > dn['warning']:
            findings.append({
                'type': 'deep_nesting', 'severity': 'warning',
                'file': filepath, 'line': line, 'function_name': name,
                'detail': f'Nesting depth is {depth} (threshold: {dn["warning"]} for warning)',
                'value': depth,
            })

        if params > lp['error']:
            findings.append({
                'type': 'long_parameter_list', 'severity': 'error',
                'file': filepath, 'line': line, 'function_name': name,
                'detail': f'Function has {params} parameters (threshold: {lp["error"]} for error)',
                'value': params,
            })
        elif params > lp['warning']:
            findings.append({
                'type': 'long_parameter_list', 'severity': 'warning',
                'file': filepath, 'line': line, 'function_name': name,
                'detail': f'Function has {params} parameters (threshold: {lp["warning"]} for warning)',
                'value': params,
            })

    return findings


def total_function_lines(filepath: str, language: str) -> int:
    """Return sum of all function body lengths in the file."""
    try:
        with open(filepath, 'r', encoding='utf-8', errors='replace') as f:
            lines = f.readlines()
    except OSError:
        return 0
    return sum(fn['body_lines'] for fn in _find_functions(lines, language))


def _find_functions(lines: list[str], language: str) -> list[dict]:
    fn_pats = FUNCTION_PATTERNS.get(language, [])
    if not fn_pats:
        return []

    found = []
    n = len(lines)

    for i, line in enumerate(lines):
        matched = any(re.search(p, line) for p in fn_pats)
        if not matched:
            continue

        fn_indent = _indent(line)
        name = _extract_name(line)
        param_count = _count_params(line, lines, i, language)

        body_lines = 0
        max_depth = 0
        j = i + 1

        while j < n:
            bline = lines[j]
            bstripped = bline.rstrip()

            if not bstripped:
                body_lines += 1
                j += 1
                continue

            b_indent = _indent(bline)
            if b_indent <= fn_indent:
                break

            body_lines += 1
            depth = b_indent // 4
            if depth > max_depth:
                max_depth = depth
            j += 1

        found.append({
            'name': name,
            'start_line': i + 1,
            'body_lines': body_lines,
            'param_count': param_count,
            'max_depth': max_depth,
        })

    return found


def _indent(line: str) -> int:
    expanded = line.expandtabs(4)
    return len(expanded) - len(expanded.lstrip())


def _extract_name(line: str) -> str:
    for pattern in [
        r'\bdef\s+(\w+)\s*\(',      # Python/Ruby
        r'\bfunction\s+(\w+)\s*\(', # JS/TS named function
        r'\bfunc\s+\(?(\w+)',        # Go
        r'\bfn\s+(\w+)',             # Rust
        r'const\s+(\w+)\s*=',       # JS arrow/const
        r'fun\s+(\w+)\s*\(',        # Kotlin
    ]:
        m = re.search(pattern, line)
        if m:
            return m.group(1)
    return '<anonymous>'


def _count_params(line: str, lines: list[str], start: int, language: str) -> int:
    paren_pos = line.find('(')
    if paren_pos < 0:
        return 0

    param_str = ''
    depth = 0
    collecting = False

    for i in range(start, min(start + 10, len(lines))):
        text = lines[i][paren_pos:] if i == start else lines[i]
        for ch in text:
            if ch == '(':
                depth += 1
                collecting = True
            elif ch == ')':
                depth -= 1
                if depth == 0 and collecting:
                    return _parse_params(param_str, language)
            elif collecting and depth > 0:
                param_str += ch

    return 0


def _parse_params(param_str: str, language: str) -> int:
    param_str = param_str.strip()
    if not param_str:
        return 0

    # Split by comma, respecting nested brackets
    depth = 0
    params = []
    current = ''
    for ch in param_str:
        if ch in '([{<':
            depth += 1
            current += ch
        elif ch in ')]}>' :
            depth -= 1
            current += ch
        elif ch == ',' and depth == 0:
            p = current.strip()
            if p:
                params.append(p)
            current = ''
        else:
            current += ch
    p = current.strip()
    if p:
        params.append(p)

    if language == 'Python':
        filtered = []
        for p in params:
            base = p.split(':')[0].split('=')[0].strip().lstrip('*')
            if base not in ('self', 'cls'):
                filtered.append(p)
        return len(filtered)

    return len(params)
